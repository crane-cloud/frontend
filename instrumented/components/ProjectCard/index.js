import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import "./ProjectCard.css";
import LineChartComponent from "../LineChart";
import getProjectMemory from "../../redux/actions/projectMemory";
import { formatMemoryMetrics } from "../../helpers/formatMetrics";
import { handleGetRequest } from "../../apis/apis.js";
import Spinner from "../Spinner";

class ProjectCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      projecMembers: [],
      fetchMembersError: "",
      currentUserRecord: {},
      fetchingProjectMembers: true,
    };
    this.getProjectMemoryMetrics = this.getProjectMemoryMetrics.bind(this);
    this.getProjectMemberz = this.getProjectMemberz.bind(this);
    this.getcurrentUserRecord = this.getcurrentUserRecord.bind(this);
    this.updateRoleValue = this.updateRoleValue.bind(this);
  }

  componentDidMount() {
    const { cardID, getProjectMemory } = this.props;
    getProjectMemory(cardID, {});
    this.getProjectMemberz();
  }

  getProjectMemoryMetrics() {
    const { cardID } = this.props;
    const { memoryMetrics } = this.props;
    const results = formatMemoryMetrics(cardID, memoryMetrics);
    return results;
  }
  getProjectMemberz() {
    const { cardID } = this.props;
    handleGetRequest(`/projects/${cardID}/users`)
      .then((response) => {
        this.setState({
          projecMembers: response.data.data.project_users,
          fetchingProjectMembers: false,
        });
        this.getcurrentUserRecord();
      })
      .catch((error) => {
        this.setState({
          fetchMembersError: "Failed to fetch project members",
          fetchingProjectMembers: false,
        });
      });
  }
  getcurrentUserRecord() {
    const { userID } = this.props;
    const { projecMembers } = this.state;
    //one object should be returned.
    this.setState({
      currentUserRecord: projecMembers?.filter(
        (item) => item.user?.id === userID
      ),
    });
  }
  updateRoleValue(string) {
    let role = string[1];
    return role.charAt(0).toUpperCase() + role.slice(1);
  }
  render() {
    const {
      name,
      description,
      cardID,
      apps_count,
      userID,
      acceptInviteCallBackModel,
      ownerId,
    } = this.props;
    const { currentUserRecord, fetchingProjectMembers } = this.state;
    const formattedMetrics = this.getProjectMemoryMetrics();
    return (
      <div className="ProjectsCard">
        <div className="appCount" title={`${apps_count} Apps in this project`}>
          <div>{apps_count}</div>
        </div>
        {/**Check if user is not owner and invitation status */}
        {userID === ownerId ||
        currentUserRecord[0]?.accepted_collaboration_invite === true ? (
          <Link
            to={{
              pathname: `/projects/${cardID}/dashboard`,
              projectData: name,
            }}
            key={cardID}
          >
            <div className="ProjectImageDiv">
              <LineChartComponent
                lineDataKey="memory"
                preview
                data={formattedMetrics}
              />
            </div>
          </Link>
        ) : (
          <div
            onClick={() => {
              acceptInviteCallBackModel(
                cardID,
                this.updateRoleValue(currentUserRecord[0].role.split("."))
              );
            }}
            className="PendingNote"
          >
            {fetchingProjectMembers ? (
              <Spinner />
            ) : (
              `Invitation to this project is pending acceptance`
            )}
          </div>
        )}
        <div className="ProjectBottomContainer">
          <div className="ProjectInfor">
            <div className="ProjectInfoSection">
              <Link
                to={
                  userID === ownerId ||
                  currentUserRecord[0]?.accepted_collaboration_invite === true
                    ? {
                        pathname: `/projects/${cardID}/dashboard`,
                        projectData: name,
                      }
                    : null
                }
                onClick={() => {
                  acceptInviteCallBackModel(
                    cardID,
                    this.updateRoleValue(currentUserRecord[0].role.split("."))
                  );
                }}
                key={cardID}
              >
                <div className="ProjectsCardName">{name}</div>
              </Link>
            </div>
            <div className="ProjectDescription">{description}</div>
          </div>
        </div>
      </div>
    );
  }
}

ProjectCard.propTypes = {
  cardID: PropTypes.string.isRequired,
  name: PropTypes.string,
  description: PropTypes.string,
  getProjectMemory: PropTypes.func.isRequired,
  memoryMetrics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

ProjectCard.defaultProps = {
  name: "",
  description: "",
};

export const mapStateToProps = (state) => {
  const { isFetchingMemory, memoryMetrics, memoryMessage } =
    state.projectMemoryReducer;
  return {
    isFetchingMemory,
    memoryMetrics,
    memoryMessage,
  };
};

const mapDispatchToProps = {
  getProjectMemory,
};

export default connect(mapStateToProps, mapDispatchToProps)(ProjectCard);
