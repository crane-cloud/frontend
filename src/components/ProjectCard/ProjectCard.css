.ProjectsCard {
  background-color: #ffffff;
  border: 1px solid #f1f1f1;
  width: 15rem;
  height: min-content;
}

.ProjectsCard:hover {
  margin: 0;
  border-color: #f7b21f;
}

.ProjectImageDiv {
  margin: 0;
  width: 100%;
  height: 9.375rem;
}

.BottomContainer {
  padding: 1rem 0.7rem;
  height: 5rem;
  display: flex;
  flex-direction: column;
  background-color: #f6fbe5;
}

.ProjectsCardName {
  width: 12rem;
  flex-grow: 1;
  color: #000;
  text-overflow: ellipsis;
  font-size: 1.5rem;
  overflow: hidden;
  white-space: nowrap;
}

.ProjectsCardName:hover {
  opacity: 0.8;
}

.ProjectInfoSection {
  display: flex;
  justify-content: space-between;
}

.ProjectDescription {
  text-overflow: ellipsis;
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}

.ProjectDropDown {
  position: relative;
}

.ProjectDropDownContent div:hover {
  background-color: #ddd;
}

.HeadingWithTooltip {
  display: inline;
}

/*///////////// Delete Model //////////////////*/

.ProjectDeleteModel {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 2;
  background-color: rgba(0, 0, 0, 0.5);
}

.ProjectDeleteModel .ModalParentWrap {
  background-color: inherit;
}

.ProjectDeleteModel .ModalChildWrap {
  z-index: 3;
}

.DeleteProjectModel {
  display: grid;
  gap: 3rem;
}

.DeleteProjectModalLowerSection {
  display: grid;
  gap: 1rem;
}

.DeleteDescription {
  cursor: default;
}

.DeleteDescription span {
  padding: 0.2rem 0.4rem;
  margin: 0;
  font-size: 90%;
  background-color: rgba(27, 31, 35, 0.1);
  border-radius: 3px;
}

.DeleteProjectModelButtons {
  display: flex;
  justify-content: space-between;
}

.DeleteProjectModelButtons .DeleteBtn {
  background-color: #ff0000;
  outline: 2px solid #ff0000;
}
.DeleteProjectModelButtons .DeleteBtn:hover {
  background-color: #fff;
  outline: 2px solid #ff0000;
  color: #ff0000;
}

/* ////////// DropDown CSS ////////// */

.DropDownImg {
  fill: #aaa;
  position: absolute;
  right: 0;
  cursor: pointer;
}

.ProjectDropDownContent {
  position: absolute;
  top: 1.5rem;
  right: 0.5rem;
  background-color: #f1f1f1;
  min-width: 6.875rem;
  z-index: 1;
}

.ProjectDropDownContent div {
  padding: 0.5rem 1rem;
  font-size: 0.8125rem;
}

.ProjectDropDownContent div:hover {
  background-color: #ddd;
}

/* ////////// Update Model ////////// */

.ModalUpdateForm {
  display: grid;
  gap: 2rem;
}

.ModalUpdateForm .ModalFormInputs {
  display: grid;
  gap: 1rem;
}

.ModalUpdateForm .ModalFormInputs .InputText {
  border: 2px solid #000000;
  color: #000;
}

.ModalUpdateForm .ModalFormInputs .InputText::placeholder {
  color: #000;
}

.ModalUpdateForm .ModalFormButtons {
  display: flex;
  justify-content: space-between;
}

.ProjectDeleteModel .ModalChildWrap {
  min-width: 21rem;
  max-width: 25rem;
}
