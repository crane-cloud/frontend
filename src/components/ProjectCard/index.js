import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import "./ProjectCard.css";
import LineChartComponent from "../LineChart";
import { formatMemoryMetrics } from "../../helpers/formatMetrics";
import { handleGetRequest } from "../../apis/apis.js";
import { ReactComponent as AlertWarning } from "../../assets/images/alert.svg";
import Spinner from "../Spinner";
import getProjectMemory from "../../redux/actions/projectMemory";
import { ReactComponent as ProjectMembers } from "../../assets/images/project-members.svg";
import { ReactComponent as Deployments } from "../../assets/images/deployments.svg";
import {formatCount} from "../../helpers/counterFormat.js"

const ProjectCard = ({
  cardID,
  name = "",
  description = "",
  apps_count,
  followers_count,
  userID,
  acceptInviteCallBackModel,
  ownerId,
  disabled,
  admin_disabled,
}) => {
  const [projectMembers, setProjectMembers] = useState([]);
  const [fetchMembersError, setFetchMembersError] = useState("");
  const [currentUserRecord, setCurrentUserRecord] = useState({});
  const [fetchingProjectMembers, setFetchingProjectMembers] = useState(true);
  const [formattedMetrics, setFormattedMetrics] = useState([]);

  const dispatch = useDispatch();
  const memoryMetrics = useSelector(
    (state) => state.projectMemoryReducer.memoryMetrics
  );

  useEffect(() => {
    dispatch(getProjectMemory(cardID, {}));
    getProjectMembers();
  }, [cardID, dispatch]);

  useEffect(() => {
    if (memoryMetrics) {
      setFormattedMetrics(formatMemoryMetrics(cardID, memoryMetrics));
    }
  }, [memoryMetrics, cardID]);

  const getProjectMembers = () => {
    handleGetRequest(`/projects/${cardID}/users`)
      .then((response) => {
        setProjectMembers(response.data.data.project_users);
        setFetchingProjectMembers(false);
        getCurrentUserRecord(response.data.data.project_users);
      })
      .catch(() => {
        setFetchMembersError("Failed to fetch project members");
        setFetchingProjectMembers(false);
      });
  };

  const getCurrentUserRecord = (members) => {
    const userRecord = members?.find((item) => item.user?.id === userID);
    setCurrentUserRecord(userRecord || {});
  };

  const updateRoleValue = (string) => {
    let role = string[1];
    return role.charAt(0).toUpperCase() + role.slice(1);
  };

  return (
    <div
      className={`ProjectsCard ${admin_disabled ? "unclickable" : ""} ${
        disabled || admin_disabled ? "opequeCard" : ""
      }`}
      title={
        disabled === true && !admin_disabled
          ? "Disabled project"
          : admin_disabled
          ? "Admin disabled this project"
          : name
      }
    >
      <div className="appCount" >
        <div className="countAndIcon">
          <ProjectMembers
            className={"membersIcon"}
            title={"Project followers"}
          />
          <span className={"memberCount"}>{formatCount(followers_count)}</span>
        </div>
        <div className="countAndIcon">
          <Deployments
            className={"membersIcon"}
            title={"Project apps"}
          />
          <span className={"memberCount"}>{formatCount(apps_count)}</span>
        </div>
      </div>
      {userID === ownerId ||
      currentUserRecord?.accepted_collaboration_invite ? (
        <Link
          to={{
            pathname: `/projects/${cardID}/dashboard`,
            projectData: name,
          }}
          className={`${admin_disabled ? "killPointerActions" : ""}`}
          key={cardID}
        >
          <div className="ProjectImageDiv">
            <LineChartComponent
              lineDataKey="memory"
              preview
              data={formattedMetrics}
            />
          </div>
        </Link>
      ) : (
        <div
          onClick={() => {
            acceptInviteCallBackModel(
              cardID,
              updateRoleValue(currentUserRecord.role.split("."))
            );
          }}
          className="PendingNote"
        >
          {fetchingProjectMembers ? (
            <Spinner />
          ) : (
            `Invitation to this project is pending acceptance`
          )}
        </div>
      )}
      <div className="ProjectBottomContainer">
        <div className="ProjectInfor">
          <div className="ProjectInfoSection">
            <Link
              to={
                userID === ownerId ||
                currentUserRecord?.accepted_collaboration_invite
                  ? {
                      pathname: `/projects/${cardID}/dashboard`,
                      projectData: name,
                    }
                  : null
              }
              className={`${admin_disabled ? "killPointerActions" : ""}`}
              onClick={() => {
                acceptInviteCallBackModel(
                  cardID,
                  updateRoleValue(currentUserRecord.role.split("."))
                );
              }}
              key={cardID}
            >
              <div className={`ProjectsCardName`}>{name}</div>
            </Link>
            {admin_disabled && (
              <div className="AlertIcon" title="Project disabled by Admin">
                <AlertWarning />
              </div>
            )}
          </div>
          <div className="ProjectDescription">{description}</div>
        </div>
      </div>
    </div>
  );
};

export default ProjectCard;
