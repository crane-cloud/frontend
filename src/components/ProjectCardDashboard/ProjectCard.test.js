/* eslint-disable no-undef */
import React from "react";
import { shallow } from "enzyme";
import ProjectCard, { mapStateToProps } from "./";

const props = {
    getProjectMemoryMetrics: jest.fn(),
    memoryMetrics: [{ slug: "slug" }],
};

describe("test the component", () => {
  it("matchs the component snapshot", () => {
    const wrapper = ProjectCard.WrappedComponent;
    const mycomponent = shallow(<wrapper {...props} />);
    expect(mycomponent).toMatchSnapshot();
  });
  it("component should match the snapshot", () => {
    const newComponent = ProjectCard.WrappedComponent;
    const wrapper = shallow(<newComponent {...props} />);
    wrapper.setProps(props);
    expect(wrapper).toBeDefined();
  });
});

describe("testing the mapstate to props ", () => {
  it(" mapstostate test", () => {
    expect(
      mapStateToProps({
        projectMemoryReducer: {
          isFetchingMemory: false,
          memoryMetrics: [],
        },
      })
    ).toEqual({
        memoryMetrics: [],
       isFetchingMemory: false,
    });
  });
});
