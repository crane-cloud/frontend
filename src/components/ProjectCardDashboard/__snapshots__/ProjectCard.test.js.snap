// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the component matchs the component snapshot 1`] = `ShallowWrapper {}`;
